### PR TITLE
8752 Added CA receiver transforms to truncate fields needed for CA

### PR DESCRIPTION
**Pull request by victor-chaparro**
This PR adds CA receiver transforms to truncate the HL7 fields needed for CA. Based on the fields listed on this ticket #8752. The following fields listed on the ticket are not being truncated:

- **MSH-3-1**: value is hardcoded and does not exceed max length
- **MSH-4-1**: value is hardcoded and does not exceed max length
- **OBR-2-2**: not sending this to CA at the moment
- **ORC-4-2**: not sending this to CA at the moment
- **PID-3-6-1**: we don't have a mapping for this field

**If you are suggesting a fix for a currently exploitable issue, please disclose the issue to the prime-reportstream team directly outside of GitHub instead of filing a PR, so we may immediately patch the affected systems before a disclosure. See [SECURITY.md/Reporting a Vulnerability](https://github.com/CDCgov/prime-reportstream/blob/master/SECURITY.md#reporting-a-vulnerability) for more information.**

Test Steps:
1. Send the attached FHIR bundle to the Reports API:
[Simple Report.postman_collection.txt](https://github.com/CDCgov/prime-reportstream/files/11165604/Simple.Report.postman_collection.txt)
This sample contains all the fields that should be truncated and you can search for them by looking for "truncatethis"

2. Verify that the request was accepted

3. Open the generated HL7 file in the local sftp folder.

4. Verify all the fields were truncated by making sure the HL7 message doesn't include the text "truncatethis"
 
5. Run integration tests and make sure they pass

## Changes
- Added CA receiver transforms to truncate HL7 fields
- Added integration test for truncating fields

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [x] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- #8752 

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #8876 

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

### CODE REVIEW

**Review by arnejduranovic — APPROVED** [review]
We discussed on a call that this work would have been simpler (and thus less error-prone) if we were able to set up a FHIR transform specific to the CA receiver. In that case, we would just need to perform the FHIR truncate operation on the FHIR fields that all the HL7 field reference, instead of translating each HL7 field separately. We should create a ticket for this work before merging, if no one disagrees.

**Comment by PatriciaPerozoUSDS** [comment]
https://github.com/CDCgov/prime-reportstream/pull/9012#pullrequestreview-1375409986

@arnejduranovic My concern about this approach is that if we do that truncation at the beginning of the pipeline by performing the FHIR truncate function, STLTs that accept un-truncated HL7 messages will experience a loss of data.

I think to make this decision we should make ourselves. Ideally I would have @jsutantio or @jakefishbein-navapbc asking STLTs that don't truncate how important that is to them . Alternatively, @brick-green or @LaFreda1 could weigh in on the STLT impact of this decision.

**Comment by PatriciaPerozoUSDS** [comment]
@victor-chaparro can you add a comment above each truncation about why its being truncated to that number of characters? I think this is something that will become confusing really quickly and I'd like it to be as easy as possible to read through the config and understand what's going on.

File: prime-router/metadata/hl7_mapping/STLTs/CA/ca-observation-result.yml [hunk 1]
```diff
@@ -7,7 +7,7 @@ elements:
 
   - name: ca-organization-performer-identifier
     resource: '%resource.performer.resolve()'
-    value: [ '%resource.name + "-"+ %resource.identifier[0].value ' ]
+    value: [ '(%resource.name + "-"+ %resource.identifier[0].value).substring(0,50) ' ]
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CA/ca-observation-result.yml:10]: https://hl7-definition.caristix.com/v2/HL7v2.5.1/Fields/OBX.23

File: prime-router/metadata/hl7_mapping/STLTs/CA/CA.yml [hunk 2]
```diff
@@ -149,3 +149,13 @@ elements:
     resource: 'Bundle.entry.resource.ofType(Patient).extension(%`rsext-patient-notes`).value.text.split(''\n'') | Bundle.entry.resource.ofType(ServiceRequest).supportingInfo.resolve()'
     condition: 'Bundle.entry.resource.ofType(Patient).extension(%`rsext-patient-notes`).exists() or Bundle.entry.resource.ofType(ServiceRequest).supportingInfo.exists()'
     schema: 'ca-patient-note'
+
+  - name: patient-assigning-authority
+    value: [ '%patientOrganization.name.substring(0,20)' ]
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CA/CA.yml:154]: https://hl7-definition.caristix.com/v2/HL7v2.5.1/Fields/PID.3.4

File: prime-router/metadata/hl7_mapping/STLTs/CA/ca-observation-request.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+constants:
+  fieldPath: /PATIENT_RESULT/ORDER_OBSERVATION(%{orderIndex})/OBR
+elements:
+
+  - name: ca-order-organization-identifier-name
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CA/ca-observation-request.yml:5]: https://hl7-definition.caristix.com/v2/HL7v2.5.1/Fields/OBR.3

File: prime-router/metadata/hl7_mapping/STLTs/CA/ca-observation-request.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    resource: '%resource.basedOn.resolve().performer.resolve()'
+    value: [ '%resource.name.substring(0,20)' ]
+    hl7Spec: [ '%{fieldPath}-3-2' ]
+
+  - name: ca-ordering-provider-identifier
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CA/ca-observation-request.yml:10]: https://hl7-definition.caristix.com/v2/HL7v2.5.1/Fields/OBR.16

File: prime-router/metadata/hl7_mapping/STLTs/CA/ca-order.yml [hunk 5]
```diff
@@ -5,4 +5,24 @@ elements:
   # if this is sent it triggers validation errors for fields we don't have
   - name: ca-ordering-facility-identifier
     value: ['']
-    hl7Spec: [ '%{hl7Order}-21-10']
\ No newline at end of file
+    hl7Spec: [ '%{hl7Order}-21-10']
+
+  - name: ca-order-organization-identifier-name
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CA/ca-order.yml:10]: https://hl7-definition.caristix.com/v2/HL7v2.5.1/Fields/ORC.3

File: prime-router/metadata/hl7_mapping/STLTs/CA/ca-order.yml
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CA/ca-order.yml:1]: I think we probably need an organizational system for the order of these fields.

My suggestion is that we put the fields in order based on their HL7 field number. its easier to read through if you see ORC-3, ORC-12, ORC-21

File: prime-router/metadata/hl7_mapping/STLTs/CA/ca-order.yml [hunk 6]
```diff
@@ -5,4 +5,24 @@ elements:
...
+    resource: '%service.performer.resolve()'
+    value: [ '%resource.name.substring(0,20)' ]
+    hl7Spec: [ '%{hl7Order}-3-2' ]
+
+  - name: ca-ordering-facility-name
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CA/ca-order.yml:15]: https://hl7-definition.caristix.com/v2/HL7v2.5.1/Fields/ORC.21

File: prime-router/metadata/hl7_mapping/STLTs/CA/ca-order.yml [hunk 7]
```diff
@@ -5,4 +5,24 @@ elements:
...
+    resource: '%service.requester.resolve().organization.resolve()'
+    value: [ '%resource.name.substring(0,50)' ]
+    hl7Spec: [ '%{hl7Order}-21-1' ]
+
+  - name: ca-order-ordering-provider-identifier
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CA/ca-order.yml:20]: https://hl7-definition.caristix.com/v2/HL7v2.5.1/Fields/ORC.12

File: prime-router/metadata/hl7_mapping/STLTs/CA/ca-patient.yml [hunk 8]
```diff
@@ -0,0 +1,20 @@
+constants:
+  hl7PIDField: '%{hl7SegmentGroup}/PID'
+elements:
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CA/ca-patient.yml:3]: https://hl7-definition.caristix.com/v2/HL7v2.5.1/Fields/PID.5

File: prime-router/metadata/hl7_mapping/STLTs/CA/ca-specimen.yml [hunk 9]
```diff
@@ -18,8 +17,21 @@ elements:
...
       identifier: '%diagnostic[%orderIndex].identifier[0].value'
-    schema: ../../common/organization-identifier
\ No newline at end of file
+    schema: ../../common/organization-identifier
+
+  #####################################
+  # California    specific transforms #
+  #####################################
+
+  - name: ca-specimen-organization-identifier-name
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CA/ca-specimen.yml:29]: https://hl7-definition.caristix.com/v2/HL7v2.5.1/Fields/SPM.2.1

File: prime-router/metadata/hl7_mapping/STLTs/CA/ca-specimen.yml [hunk 10]
```diff
@@ -18,8 +17,21 @@ elements:
...
+    resource: '%performerOrganization'
+    value: [ '%resource.name.substring(0,20)' ]
+    hl7Spec: [ '%{hl7SpecimenFieldPath}-2-1-2' ]
+
+  - name: ca-specimen-organization-identifier-name-2
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CA/ca-specimen.yml:34]: https://hl7-definition.caristix.com/v2/HL7v2.5.1/Fields/SPM.2.2

**Review by PatriciaPerozoUSDS — APPROVED** [review]
Once you add the comments so that its clear how we got to that number this is good to go

File: prime-router/metadata/hl7_mapping/STLTs/CA/ca-specimen.yml [hunk 11]
```diff
@@ -18,8 +17,22 @@ elements:
   # Same as above
   - name: ca-specimen-organization-identifier-2
     resource: '%performerOrganization'
-    debug: true
     constants:
       organizationIdFieldPath: '%{hl7SpecimenFieldPath}-2-2'
       identifier: '%diagnostic[%orderIndex].identifier[0].value'
```
Comment (PatriciaPerozoUSDS) [inline prime-router/metadata/hl7_mapping/STLTs/CA/ca-specimen.yml:22]: I don't think I caught this in the code review - I thought we were pulling from the specimen and not the diagnostic report now that SR fixed their unique problem?

**Pull request merged**